src/libcore/ffi.rs

```diff
@@ -111,22 +125,117 @@ struct VaListImpl {
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
-struct VaListImpl {
+#[lang = "va_list"]
+pub struct VaListImpl<'f> {
     gp_offset: i32,
     fp_offset: i32,
     overflow_arg_area: *mut c_void,
     reg_save_area: *mut c_void,
+    _marker: PhantomData<&'f c_void>,
 }
 
-/// A wrapper for a `va_list`
-#[lang = "va_list"]
+/// asm.js ABI implementation of a `va_list`.
+// asm.js uses the PNaCl ABI, which specifies that a `va_list` is
+// an array of 4 32-bit integers, according to the old PNaCl docs at
+// https://web.archive.org/web/20130518054430/https://www.chromium.org/nativeclient/pnacl/bitcode-abi#TOC-Derived-Types
+#[cfg(all(target_arch = "asmjs", not(windows)))]
+#[repr(C)]
 #[derive(Debug)]
 #[unstable(feature = "c_variadic",
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
+#[lang = "va_list"]
+pub struct VaListImpl<'f> {
+    inner: [i32; 4],
```

Comment:
If Rust code doesn't need to touch this, I'd just use `MaybeUninit<[i32; 4]>`.

Reply:
Rust doesn't touch this `inner` directly.

`va_start` and `va_copy` currently only set the first element of  the array, so would `[MaybeUninit<i32>; 4]` be better?

Reply:
Adding `MaybeUninit` now gives me a warning:
```
warning: `extern` block uses type `std::mem::MaybeUninit<i32>` which is not FFI-safe: this union has unspecified layout
  --> $DIR/variadic-ffi.rs:9:51
   |
LL |     fn rust_valist_interesting_average(_: u64, _: VaList) -> f64;
   |                                                   ^^^^^^
   |
   = note: #[warn(improper_ctypes)] on by default
   = help: consider adding a #[repr(C)] attribute to this union
```

Reply:
cc @RalfJung Should we consider `MaybeUninit` FFI-safe?

Reply:
It's FFI-safe in-memory, but I don't know about the by-value situation. Cc @rkruppe 

Last time this came up being doubted there would be any use for passing `MaybeUninit` by-value. That was in https://github.com/rust-lang/rfcs/pull/2645.

Reply:
I think that in this case `VaList` contains a reference to the `struct` containing a `MaybeUninit`, instead of directly containing it.

Reply:
Oh, but then we just need it to be the same in memory, right? That's easy.

(I never know what people mean when they ask about stuff being FFI-safe...)

Reply:
@RalfJung Sorry, by "should we consider" I was referring to the `improper_ctypes` lint, as seen above.

Reply:
I'm afraid I don't know the implications of that :/ @gnzlb and @rkruppe would be the people to ask, IIRC.

Reply:
improper_ctypes doesn't currently have a distinction between "ok behind a pointer" and "ok to pass by value". That would be useful for several cases, including ZSTs arrays. This specific case is likely to be solved sooner by MaybeUninit becoming repr(transparent) though.

Reply:
@RalfJung 

> Last time this came up people doubted there would be any use for passing MaybeUninit by-value. 

What was argued in the RFC is that using `MaybeUninit` on FFI does not require ABI-compatibility (which is what `repr(transparent)` enforces); it only requires layout-compatibility. 

AFAICT, `MaybeUninit<[i32; 4]>` is layout-compatible with `[i32; 4]`, and if we don't guarantee that, we should probably do so. Since `[i32; 4]` is `repr(C)` (for better or worse - #36464), using `MaybeUninit<[i32; 4]>` here on C FFI should be guaranteed to work as expected, independently of whether `MaybeUninit` is `repr(transparent)` or not.

Reply:
> What was argued in the RFC is that using MaybeUninit on FFI does not require ABI-compatibility (which is what repr(transparent) enforces)

Yes, that's what I referred to with by-value.

> AFAICT, MaybeUninit<[i32; 4]> is layout-compatible with [i32; 4], and if we don't guarantee that, we should probably do so. Since [i32; 4] is repr(C) (for better or worse - #36464), using MaybeUninit<[i32; 4]> here on C FFI should be guaranteed to work as expected, independently of whether MaybeUninit is repr(transparent) or not.

We are in agreement here.

Reply:
Is it possible to `#[allow(improper_ctypes)]` on a per-type basis (e.g. for `VaList`) ? If so, maybe the simplest thing would be to just add the attribute to `VaList` and a comment explaining why are we ignoring the warning here, and that doing so is "ok" for now. 

Improving the `improper_ctypes` lint to handle these, or making `MaybeUninit` `repr(transparent)` would fix this issue, but can be done later.

Reply:
I disabled the lint for just that test in `variadic-ffi.rs`, all asm.js tests pass now.